### PR TITLE
Feature: New option to limit routes to certain API tokens

### PR DESCRIPTION
**Pull request by hipio**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -104,6 +105,36 @@ You can create API tokens through code within classes that are `ContainerAware`.
...
+
+    api_token_security:
+      enforce_path_definition: true
+      role_prefix: ~
+
+These two values have to be declared, but can be set to null (`~`) if one desires to do so. 
+The default values are as following.
+
+    enforce_path_definition: false
+    pre_role_value: API_TOKEN_
```
Comment (Potherca) [inline README.md]: Shouldn't `pre_role_value` be `role_prefix`?

File: src/DependencyInjection/Configuration.php [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+        $rootNode->children()
+            /* Defining enforce_path_definition setting */
+            ->booleanNode('enforce_path_definition')->treatNullLike(false)->isRequired()
+            ->info('Should we enforce that API paths should be defined before being accessible to valid API tokens?')
+            ->end()
+            /* End defining enforce_path_definition setting */
+
+            /* Defining role_prefix setting */
+            ->scalarNode('role_prefix')->defaultValue('API_TOKEN_')->treatNullLike('API_TOKEN_')->isRequired()
+            ->info('What is the value should prepend all our security.yml access_control route roles?')
```
Comment (Potherca) [inline src/DependencyInjection/Configuration.php]: Change to:

> Which value should prepend the security.yml access_control route roles?

File: src/Security/AccessVoter.php [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+    /** @var ContainerInterface */
+    private $doctrine;
+
+    /** @var string */
+    private $preRoleValue;
+
+    /**
+     * The constructor for the AccessVoter class.
+     *
+     * @param Registry $container
```
Comment (Potherca) [inline src/Security/AccessVoter.php]: Missing param declaration.

File: src/Security/AccessVoter.php [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+
+        if ($apiToken === null) {
+            return false;
+        }
+
+        if ($subject instanceof Request === false) {
+            return false;
+        }
+
+        return true;
```
Comment (Potherca) [inline src/Security/AccessVoter.php]: Function can be simplified:

```
    $apiToken = $this->validAPIToken($attribute);

    return $apiToken !== null && $subject instanceof Request;
```

File: src/Security/AccessVoter.php [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+use Symfony\Component\Security\Core\Authorization\Voter\Voter;
+
+class AccessVoter extends Voter
+{
+
+    /** @var ContainerInterface */
+    private $doctrine;
+
+    /** @var string */
+    private $preRoleValue;
```
Comment (Potherca) [inline src/Security/AccessVoter.php]: `prefix`?

File: src/Security/AccessVoter.php [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+     *
+     * @return mixed
+     */
+    private function convertAttributeToTokenname($attribute)
+    {
+        $tokenName = str_replace($this->preRoleValue, '', $attribute);
+
+        return $tokenName;
+    }
+}
```
Comment (Potherca) [inline src/Security/AccessVoter.php]: Missing trailing linebreak.

File: src/SecurityAPIBundle.php [hunk 7]
```diff
@@ -30,6 +31,11 @@ public function build(ContainerBuilder $container)
         $this->addRegisterMappingsPass($container);
     }
 
+    public function getContainerExtension()
+    {
+        return new ConfigurationExtension();
```
Comment (Potherca) [inline src/SecurityAPIBundle.php]: Is there a reason this extension is not injected into the class?
Also, as this is not a pure getter, rename to `createContainerExtension` instead.

File: README.md (hunk 1 above)
Comment (hipio) [inline README.md]: Yes. Last minute changes are a bad idea.

File: src/DependencyInjection/Configuration.php (hunk 2 above)
Comment (hipio) [inline src/DependencyInjection/Configuration.php]: 👍

File: src/Security/AccessVoter.php (hunk 5 above)
Comment (hipio) [inline src/Security/AccessVoter.php]: 👍

File: src/Security/AccessVoter.php (hunk 3 above)
Comment (hipio) [inline src/Security/AccessVoter.php]: 👍

File: src/Entity/APIToken.php [hunk 8]
```diff
@@ -116,6 +116,6 @@ public function setCreatedAt($createdAt)
      */
     public function __toString()
     {
-        return (string) $this->getName();
+        return (string)$this->getName();
```
Comment (arjank) [inline src/Entity/APIToken.php]: If I'm correct, PSR2 indicates that there should be a space between `(string)` and `$this->getName()`. Also, if that's the only change in a file, it's most likely not worth committing the file.

File: src/Security/AccessVoter.php (hunk 4 above)
Comment (hipio) [inline src/Security/AccessVoter.php]: Yes, but does that make it more readable? I think the current way is a bit easier to read.

File: src/Security/AccessVoter.php (hunk 6 above)
Comment (hipio) [inline src/Security/AccessVoter.php]: Fixed in last commit.

File: src/SecurityAPIBundle.php (hunk 7 above)
Comment (hipio) [inline src/SecurityAPIBundle.php]: This is how Symfony does it. Not much I can change about it.

File: src/Security/AccessVoter.php (hunk 4 above)
Comment (Potherca) [inline src/Security/AccessVoter.php]: Having two `if`s and three return points doesn't really scream "readability", IMHO.

Comment (hipio) [inline src/Security/AccessVoter.php]: True, but if I was reading unknown code I would prefer this over the short statement. Symfony themselves also implements the `supports` function like this (https://symfony.com/doc/current/security/voters.html#creating-the-custom-voter).

I know that this is a logic fallacy (appealing to authority), but I really do think this is nicer to read. Should we ask for a second opinion?

Comment (Potherca) [inline src/Security/AccessVoter.php]: Alternative suggestion:

```
    $supports = false;

    if ($apiToken !== null && $subject instanceof Request) {
        $supports = true;
    }

    return $supports;
```

Comment (hipio) [inline src/Security/AccessVoter.php]: Well, I rather use the original suggestion than this one. 😛

File: src/Entity/APIToken.php (hunk 8 above)
Comment (hipio) [inline src/Entity/APIToken.php]: PHPCS insists that it should be wedged together.

Comment (arjank) [inline src/Entity/APIToken.php]: Then I'm probably mistaken.

Comment (hipio) [inline src/Entity/APIToken.php]: Odd, I'll fix it.

**Pull request merged**